README.md
# raiden-contracts

## Prerequisites


-  Python 3.6
-  https://pip.pypa.io/en/stable/

## Installation

```
make install
```

## Deployment on a testnet


```
# Following calls are equivalent

python -m deploy

python -m deploy \
       --rpc-provider http://127.0.0.1:8545 \
       --json raiden_contracts/data/contracts.json
       --owner 0x5601Ea8445A5d96EEeBF89A67C4199FbB7a43Fbb \
       --wait 300 \
       --token-name CustomToken --token-symbol TKN \
       --supply 10000000 --token-decimals 18

# Provide a custom deployed token
python -m deploy --token-address <TOKEN_ADDRESS>

```

## Development

We use `populus` for development. At the moment, this library is incompatible with recent libraries that act as dependencies. Therefore, we have to use different environments here.

```bash
pip install -r requirements-dev.txt
```

After you are finished, to uninstall `populus` and install the dependencies for the `raiden-contracts` package, you can do:

```
pip freeze  | xargs pip uninstall -y
pip install -r requirements.txt

```

### Compile the contracts

```
python setup.py build
```

### Testing

```
# tests
pytest
pytest -p no:warnings -s
pytest raiden_contracts/tests/test_token_network.py -p no:warnings -s

# Recommended for speed:
pip install pytest-xdist==1.17.1
pytest -p no:warnings -s -n NUM_OF_CPUs

```


CHANGELOG.md
# Changelog

Documents changes that result in:
- smart contracts redeployment, especially ABI changes
- API changes in the package (externally used constants, externally used utilities and scripts)
- important bug fixes between releases

## [0.50.2]
- Reduce dependencies and dependency updates
- Drop python3.7 support

## [0.50.0]

- [#1580](https://github.com/raiden-network/raiden-contracts/issues/1580) Establish better naming convention for time related values
- [#1550](https://github.com/raiden-network/raiden-contracts/issues/1550) Replace settle timeout range with one constant value
- [#1582](https://github.com/raiden-network/raiden-contracts/issues/1582) Move settle timeouts from channels to token networks
- [#1565](https://github.com/raiden-network/raiden-contracts/issues/1565) Switch from block numbers to dates/time values
- [#1612](https://github.com/raiden-network/raiden-contracts/issues/1612) Add deployment option for user deposit withdraw timeout
- [#1608](https://github.com/raiden-network/raiden-contracts/issues/1608) Extend documentation for Arbitrum gas prices
- [#1623](https://github.com/raiden-network/raiden-contracts/pull/1623) Add API endpoint to verify Arbitrum One deployments

## [0.40.3]

- [#1602](https://github.com/raiden-network/raiden-contracts/pull/1602) Upgrade and unpin `web3` dependency

## [0.40.2]

- [#1584](https://github.com/raiden-network/raiden-contracts/pull/1584) Upgrade `web3` and `eth-tester` dependencies

## [0.40.1]

- [#1576](https://github.com/raiden-network/raiden-contracts/pull/1576) Add missing `mypy_extensions` requirement

## [0.40.0]

- [#1517](https://github.com/raiden-network/raiden-contracts/pull/1517), [#1522](https://github.com/raiden-network/raiden-contracts/pull/1522) Upgrade to solidity 0.8 series
- [#1519](https://github.com/raiden-network/raiden-contracts/pull/1519) Use `block.chainid` in contracts instead of a deploy time value
  - This changes the ABI of the `TokenNetworkRegistry` and `TokenNetwork` contract constructors

## [0.39.0]

- [#81](https://github.com/raiden-network/raiden-contracts/issues/81) Add error messages to all contracts
- [#1182](https://github.com/raiden-network/raiden-contracts/issues/1182) Add support for reusing the `ServiceRegistry` in the deployment scripts
- [#1383](https://github.com/raiden-network/raiden-contracts/issues/1383) Add `TokenNetwork.cooperativeSettle`, which settles the channel in a single transaction when the participants cooperate
- [#1436](https://github.com/raiden-network/raiden-contracts/issues/1436) Add `TokenNetwork.openChannelWithDeposit`, which combines opening and depositing to a channel
- [#1481](https://github.com/raiden-network/raiden-contracts/issues/1481) Allow removing the limits after deployments
- [#1482](https://github.com/raiden-network/raiden-contracts/issues/1482) Add `UserDeposit.withdrawToBeneficiary`, which allows withdrawing to a beneficiary
- [#1494](https://github.com/raiden-network/raiden-contracts/pull/1494) Add both participants addresses to the `TokenNetwork.ChannelSettled` event
- [#1496](https://github.com/raiden-network/raiden-contracts/pull/1496) Enable contract optimizations

## [0.38.0]
- [#1462](https://github.com/raiden-network/raiden-contracts/pull/1462) TokenNetwork: allow updating channel state anytime before a channel is settled
- [#1464](https://github.com/raiden-network/raiden-contracts/pull/1464) Use RDN as default token for service payments

## [0.37.6]
- [#1456](https://github.com/raiden-network/raiden-contracts/pull/1456) Add Python 3.9 support

## [0.37.5]
- Add missing deployment files for the unstable deployment done in 0.37.2.

## [0.37.3]

- [#1443](https://github.com/raiden-network/raiden-contracts/pull/1443) Remove `semver` dependency, which was causing version conflicts for other packages using `raiden-contracts`

## [0.37.2]

- [#1439](https://github.com/raiden-network/raiden-contracts/pull/1439) Provide additional unstable (development) deployment on goerli testnet
- [#1430](https://github.com/raiden-network/raiden-contracts/pull/1430) Allow reusing the ServiceRegistry when deploying contracts

## [0.37.0]

- New deployments for both mainnet and testnets (goerli, rinkeby, ropsten)
- Updated disclaimer

## [0.37.0-b3]

- Fix `CONTRACTS_VERSION`, which was outdated in 0.37.0-b2.

## [0.37.0-b2]

- [#1376](https://github.com/raiden-network/raiden-contracts/pull/1376) Support verification of TokenNetworks
- Test release of contracts on mainnet for Brainbot-internal testing. Do not use for anything else! The deprecation switch will be triggered soon!

## [0.37.0-beta]

- Practice deployment for next mainnet release

## [0.36.2]

- [#1372](https://github.com/raiden-network/raiden-contracts/pull/1372) Rename network to chain for consistency
- [#1371](https://github.com/raiden-network/raiden-contracts/pull/1371) Update `web3` and improve typing
- [#1363](https://github.com/raiden-network/raiden-contracts/pull/1363) Update `solc` to 0.6.3

## [0.36.1]

- [#1339](https://github.com/raiden-network/raiden-contracts/pull/1339) Update `web3` and related dependencies. This also enables an Istanbul compatible test chain.
- [#1361](https://github.com/raiden-network/raiden-contracts/pull/1361) Require `--service-registry-controller` during service contracts deployment.
- [#1360](https://github.com/raiden-network/raiden-contracts/pull/1360) Add Python 3.8 support

## [0.36.0]

- [#1355](https://github.com/raiden-network/raiden-contracts/pull/1355) update solc version to 0.5.16
- allow unlimited number of TokenNetworks in testnet deployment

## [0.35.0]

- No code changes, just different parameters for testnet deployments than in
  0.34. Settings are the same as in the 0.33 deployments, again.

## [0.34.0]

- [#1318](https://github.com/raiden-network/raiden-contracts/pull/1318) add `make install-dev`.
- [#1350](https://github.com/raiden-network/raiden-contracts/pull/1350) remove support for deploying unlimited contracts


deploy/__main__.py
'''
A simple Python script to deploy contracts.
'''
import click
from populus import Project
from eth_utils import (
    is_address,
    to_checksum_address,
)
from utils.utils import (
    check_succesful_tx
)


@click.command()
@click.option(
    '--chain',
    default='ropsten',
    help='Chain to deploy on: kovan | ropsten | rinkeby | tester | privtest'
)
@click.option(
    '--owner',
    help='Contracts owner, default: web3.eth.accounts[0]'
)
@click.option(
    '--supply',
    default=10000000,
    help='Token contract supply (number of total issued tokens).'
)
@click.option(
    '--token-name',
    default='CustomToken',
    help='Token contract name.'
)
@click.option(
    '--token-decimals',
    default=18,
    help='Token contract number of decimals.'
)
@click.option(
    '--token-symbol',
    default='TKN',
    help='Token contract symbol.'
)
@click.option(
    '--token-address',
    help='Already deployed token address.'
)
def main(**kwargs):
    project = Project()

    chain_name = kwargs['chain']
    owner = kwargs['owner']
    supply = kwargs['supply']
    token_name = kwargs['token_name']
    token_decimals = kwargs['token_decimals']
    token_symbol = kwargs['token_symbol']
    token_address = kwargs['token_address']
    supply *= 10**(token_decimals)
    txn_wait = 300

    print('''Make sure {} chain is running, you can connect to it and it is synced,
          or you'll get timeout'''.format(chain_name))


    with project.get_chain(chain_name) as chain:
        web3 = chain.web3
        print('Web3 provider is', web3.providers[0])

        owner = owner or web3.eth.accounts[0]
        assert owner and is_address(owner), 'Invalid owner provided.'
        owner = to_checksum_address(owner)
        print('Owner is', owner)
        assert web3.eth.getBalance(owner) > 0, 'Account with insuficient funds.'

        def deploy_contract(contract_name, args=[]):
            contract = chain.provider.get_contract_factory(contract_name)
            txhash = contract.deploy(
                args=args,
                transaction={'from': owner}
            )
            receipt = check_succesful_tx(chain.web3, txhash, txn_wait)
            return receipt['contractAddress']

        token = chain.provider.get_contract_factory('CustomToken')

        if not token_address:
            txhash = token.deploy(
                args=[supply, token_decimals, token_name, token_symbol],
                transaction={'from': owner}
            )
            receipt = check_succesful_tx(chain.web3, txhash, txn_wait)
            token_address = receipt['contractAddress']

        assert token_address and is_address(token_address)
        token_address = to_checksum_address(token_address)
        print(token_name, 'address is', token_address)

        secret_registry_address = deploy_contract('SecretRegistry')
        print('SecretRegistry address:', secret_registry_address)

        token_network_registry_address = deploy_contract('TokenNetworkRegistry', [secret_registry_address])
        print('TokenNetworkRegistry address:', token_network_registry_address)

        token_network_address = deploy_contract('TokenNetwork', [token_address, secret_registry_address])
        print('TokenNetwork address:', token_network_address)


if __name__ == '__main__':
    main()
